### ISSUE TITLE
tf object detection api with cross gpu batch norm: LossTensor is inf or nan

### ISSUE THREAD

**Issue by MrWanter** [tensorflow/models#4744]
Hi! I tried implement the cross gpu batchnorm ref for [cross gpu batchnorm](https://github.com/tensorpack/tensorpack/blob/07783edb998cec3ec91c4312b39bd754cf9ececa/tensorpack/models/batch_norm.py#L198-L253), I can run it correctly on single machine, and get no error, but when deployed to distributed training, which was also tested successfully with TF object detection api, it raises:
```
2018-07-11 21:42:21.794006: W tensorflow/stream_executor/cuda/cuda_dnn.cc:3797] 
INFO:tensorflow:Recording summary at step 0.
INFO:tensorflow:Recording summary at step 0.
INFO:tensorflow:global step 0: loss = 8.5886 (18.891 sec/step)
INFO:tensorflow:global step 0: loss = 8.5886 (18.891 sec/step)
INFO:tensorflow:Recording summary at step 0.
INFO:tensorflow:Recording summary at step 0.
INFO:tensorflow:global step 0: loss = 8.2428 (7.556 sec/step)
INFO:tensorflow:global step 0: loss = 8.2428 (7.556 sec/step)
INFO:tensorflow:global step 0: loss = 8.7922 (6.760 sec/step)
INFO:tensorflow:global step 0: loss = 8.7922 (6.760 sec/step)
INFO:tensorflow:global step 0: loss = 8.3253 (6.729 sec/step)
INFO:tensorflow:global step 0: loss = 8.3253 (6.729 sec/step)
INFO:tensorflow:Error reported to Coordinator: <class 'tensorflow.python.framework.errors_impl.InvalidArgumentErro
r'>, LossTensor is inf or nan. : Tensor had NaN values
         [[Node: CheckNumerics = CheckNumerics[T=DT_FLOAT, message="LossTensor is inf or nan.", _device="/job:mast
er/replica:0/task:0/device:CPU:0"](total_loss)]]
         [[Node: clip_grads/clip_by_norm_428/truediv_S17125 = _Recv[client_terminated=false, recv_device="/job:ps/
replica:0/task:2/device:CPU:0", send_device="/job:master/replica:0/task:0/device:CPU:0", send_device_incarnation=6
739474025618584086, tensor_name="edge_392315_clip_grads/clip_by_norm_428/truediv", tensor_type=DT_FLOAT, _device="
/job:ps/replica:0/task:2/device:CPU:0"]()]]

Caused by op u'CheckNumerics', defined at:
  File "train.py", line 238, in <module>
    tf.app.run()
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/platform/app.py", 
line 126, in run
    _sys.exit(main(argv))
  File "train.py", line 234, in main
    graph_hook_fn=graph_rewriter_fn)
  File "/home/kattk/prj/tf-object-api-orig-k80/object_detection/trainer.py", line 311, in trai
n
    total_loss = tf.check_numerics(total_loss, 'LossTensor is inf or nan.')
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/ops/gen_array_ops.
py", line 737, in check_numerics
    "CheckNumerics", tensor=tensor, message=message, name=name)
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/op_def_l
ibrary.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/ops.py",
 line 3392, in create_op
    op_def=op_def)
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/ops.py",
 line 1718, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): LossTensor is inf or nan. : Tensor had NaN values
         [[Node: CheckNumerics = CheckNumerics[T=DT_FLOAT, message="LossTensor is inf or nan.", _device="/job:mast
er/replica:0/task:0/device:CPU:0"](total_loss)]]
         [[Node: clip_grads/clip_by_norm_428/truediv_S17125 = _Recv[client_terminated=false, recv_device="/job:ps/
replica:0/task:2/device:CPU:0", send_device="/job:master/replica:0/task:0/device:CPU:0", send_device_incarnation=6
739474025618584086, tensor_name="edge_392315_clip_grads/clip_by_norm_428/truediv", tensor_type=DT_FLOAT, _device="
/job:ps/replica:0/task:2/device:CPU:0"]()]]

INFO:tensorflow:Error reported to Coordinator: <class 'tensorflow.python.framework.errors_impl.InvalidArgumentErro
r'>, LossTensor is inf or nan. : Tensor had NaN values
         [[Node: CheckNumerics = CheckNumerics[T=DT_FLOAT, message="LossTensor is inf or nan.", _device="/job:maser/replica:0/task:0/device:CPU:0"](total_loss)]]
         [[Node: clip_grads/clip_by_norm_428/truediv_S17125 = _Recv[client_terminated=false, recv_device="/job:ps/
replica:0/task:2/device:CPU:0", send_device="/job:master/replica:0/task:0/device:CPU:0", send_device_incarnation=6
739474025618584086, tensor_name="edge_392315_clip_grads/clip_by_norm_428/truediv", tensor_type=DT_FLOAT, _device=$
/job:ps/replica:0/task:2/device:CPU:0"]()]]

Caused by op u'CheckNumerics', defined at:
  File "train.py", line 238, in <module>
    tf.app.run()
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/platform/app.py", 
line 126, in run
    _sys.exit(main(argv))
  File "train.py", line 234, in main
    graph_hook_fn=graph_rewriter_fn)
  File "/home/kattk/prj/tf-object-api-orig-k80/object_detection/trainer.py", line 311, in tra$
n
    total_loss = tf.check_numerics(total_loss, 'LossTensor is inf or nan.')
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/ops/gen_array_ops$
py", line 737, in check_numerics
    "CheckNumerics", tensor=tensor, message=message, name=name)
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/op_def_$
ibrary.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/ops.py"$
 line 3392, in create_op
    op_def=op_def)
  File "/home/kattk/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/ops.py"$
 line 1718, in __init__

```
I cant find any clue for the error reported, can someone help?

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by harshini-gadige** [comment]
Closing as this has been answered and in "awaiting response" status for more than a week.  Please add comments if any, we will reopen it. Thanks !
